File: lib/models/product.dart
class Product {
  final String id;
  final String userId;
  final String name;
  final String description;
  final double price;
  final String category;
  final String condition;
  final String status;
  final String? imageUrl;
  final int viewCount;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? sellerName;
  final String? sellerUsername;

  Product({
    required this.id,
    required this.userId,
    required this.name,
    required this.description,
    required this.price,
    required this.category,
    required this.condition,
    required this.status,
    this.imageUrl,
    required this.viewCount,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
    this.sellerName,
    this.sellerUsername,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] ?? '',
      userId: json['user_id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      price: (json['price'] ?? 0).toDouble(),
      category: json['category'] ?? '',
      condition: json['condition'] ?? '',
      status: json['status'] ?? '',
      imageUrl: json['image_url'],
      viewCount: json['view_count'] ?? 0,
      isActive: json['is_active'] ?? true,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : DateTime.now(),
      sellerName: json['seller_name'],
      sellerUsername: json['seller_username'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'name': name,
      'description': description,
      'price': price,
      'category': category,
      'condition': condition,
      'status': status,
      'image_url': imageUrl,
      'view_count': viewCount,
      'is_active': isActive,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'seller_name': sellerName,
      'seller_username': sellerUsername,
    };
  }

  String get formattedPrice {
    return 'Rp ${price.toStringAsFixed(0).replaceAllMapped(
          RegExp(r'(\d{1,3})(?=(\d{3})+(?!\d))'),
          (Match m) => '${m[1]}.',
        )}';
  }

  String get categoryDisplay {
    switch (category.toLowerCase()) {
      case 'gaming':
        return 'Gaming';
      case 'laptop':
        return 'Laptop';
      case 'phone':
        return 'Phone';
      case 'audio':
        return 'Audio';
      case 'camera':
        return 'Camera';
      case 'watch':
        return 'Watch';
      case 'tablet':
        return 'Tablet';
      case 'accessories':
        return 'Accessories';
      default:
        return category;
    }
  }

  String get conditionDisplay {
    switch (condition.toLowerCase()) {
      case 'new':
        return 'New';
      case 'like_new':
        return 'Like New';
      case 'good':
        return 'Good';
      case 'fair':
        return 'Fair';
      default:
        return condition;
    }
  }

  String get statusDisplay {
    switch (status.toLowerCase()) {
      case 'available':
        return 'Available';
      case 'sold':
        return 'Sold';
      case 'reserved':
        return 'Reserved';
      default:
        return status;
    }
  }
}

File: backend/models/product.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Product model - tabel products
type Product struct {
	ID          string         `gorm:"type:char(36);primaryKey" json:"id"`
	UserID      string         `gorm:"type:char(36);not null;index" json:"user_id"`
	Name        string         `gorm:"type:varchar(255);not null" json:"name"`
	Description string         `gorm:"type:text" json:"description"`
	Price       float64        `gorm:"type:decimal(12,2);not null" json:"price"`
	Category    string         `gorm:"type:varchar(50);index" json:"category"`
	Condition   string         `gorm:"type:varchar(20)" json:"condition"` // new, like_new, good, fair
	Status      string         `gorm:"type:varchar(20);default:'available';index" json:"status"` // available, sold, reserved
	ImageURL    string         `gorm:"type:varchar(500)" json:"image_url"`
	ViewCount   int            `gorm:"default:0" json:"view_count"`
	IsActive    bool           `gorm:"default:true" json:"is_active"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`

	// Relation
	User User `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

// TableName override nama tabel
func (Product) TableName() string {
	return "products"
}

// ProductResponse - response dengan seller info
type ProductResponse struct {
	ID          string    `json:"id"`
	UserID      string    `json:"user_id"`
	Name        string    `json:"name"`
	Description string    `json:"description"`
	Price       float64   `json:"price"`
	Category    string    `json:"category"`
	Condition   string    `json:"condition"`
	Status      string    `json:"status"`
	ImageURL    string    `json:"image_url"`
	ViewCount   int       `json:"view_count"`
	IsActive    bool      `json:"is_active"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	
	// Seller info
	SellerName     string `json:"seller_name,omitempty"`
	SellerUsername string `json:"seller_username,omitempty"`
}

// ToResponse convert Product ke ProductResponse
func (p *Product) ToResponse() ProductResponse {
	return ProductResponse{
		ID:             p.ID,
		UserID:         p.UserID,
		Name:           p.Name,
		Description:    p.Description,
		Price:          p.Price,
		Category:       p.Category,
		Condition:      p.Condition,
		Status:         p.Status,
		ImageURL:       p.ImageURL,
		ViewCount:      p.ViewCount,
		IsActive:       p.IsActive,
		CreatedAt:      p.CreatedAt,
		UpdatedAt:      p.UpdatedAt,
		SellerName:     p.User.FullName,
		SellerUsername: p.User.Username,
	}
}
